File: 18-4sum/4sum.c
#include <stdlib.h>

int cmp(const void *a, const void *b) {
    long x = *(int *)a;
    long y = *(int *)b;
    return (x > y) - (x < y);
}

int** fourSum(int* nums, int numsSize, int target,
              int* returnSize, int** returnColumnSizes) {

    qsort(nums, numsSize, sizeof(int), cmp);

    int capacity = 1000;
    int **result = (int **)malloc(capacity * sizeof(int *));
    *returnColumnSizes = (int *)malloc(capacity * sizeof(int));

    *returnSize = 0;

    for (int i = 0; i < numsSize - 3; i++) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;

        for (int j = i + 1; j < numsSize - 2; j++) {
            if (j > i + 1 && nums[j] == nums[j - 1]) continue;

            int left = j + 1, right = numsSize - 1;

            while (left < right) {
                long sum = (long)nums[i] + nums[j] + nums[left] + nums[right];

                if (sum == target) {
                    int *quad = (int *)malloc(4 * sizeof(int));
                    quad[0] = nums[i];
                    quad[1] = nums[j];
                    quad[2] = nums[left];
                    quad[3] = nums[right];

                    result[*returnSize] = quad;
                    (*returnColumnSizes)[*returnSize] = 4;
                    (*returnSize)++;

                    left++;
                    right--;

                    while (left < right && nums[left] == nums[left - 1]) left++;
                    while (left < right && nums[right] == nums[right + 1]) right--;
                }
                else if (sum < target) {
                    left++;
                }
                else {
                    right--;
                }
            }
        }
    }

    return result;
}

File: 18-4sum/4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> arr=new ArrayList<>();
        int n=nums.length;
        Arrays.sort(nums);
          for(int i=0;i<n;i++){
         if(i>0&&nums[i]==nums[i-1]){
            continue;
         }
         for(int j=i+1;j<n;j++){
              if(j>i+1&&nums[j]==nums[j-1]){
            continue;
         }
         int left=j+1;
         int right=n-1;
         
         while(left<right){
         long sum=(long)nums[i]+nums[j]+nums[left]+nums[right];
            List<Integer> z=new ArrayList<>();
            if(sum==target){
            z.add(nums[i]);
            z.add(nums[j]);
            z.add(nums[left]);
            z.add(nums[right]);
            arr.add(z);
            while(left<right&&nums[left]==nums[left+1]){
                left++;
            }
            while(left<right&&nums[right]==nums[right-1]){
                right--;
            }
            left++;
            right--;
            }else if(sum<target){
                left++;
            }else{
                right--;
            }
         }
         }
     }
     return arr;
    }
}